File: src/main/java/no/nav/samordning/hendelser/metrics/AppMetrics.java
package no.nav.samordning.hendelser.metrics;

import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.Gauge;
import io.micrometer.core.instrument.MeterRegistry;
import no.nav.samordning.hendelser.database.Database;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.stereotype.Component;

import java.util.*;

@Component
public class AppMetrics {

    private static final Logger LOG = LoggerFactory.getLogger(AppMetrics.class);

    @Autowired
    private Database database;

    private Number totalAntallHendelser = 0;

    private Map<String, Counter> hendelserLestCounterList = new HashMap<>();

    public AppMetrics(MeterRegistry registry) {
        Gauge.builder("samordning_hendelser_total", () -> totalAntallHendelser).register(registry);
        hendelserLestCounterList.put("3010", Counter.builder("samordning_hendelser_lest").tag("tpnr", "3010").register(registry));
    }

    @Bean
    public void totalHendelserCount() {
        TimerTask counterTask = new TimerTask() {
            @Override
            public void run() {
                totalAntallHendelser = Integer.parseInt(database.getTotalHendelser());
            }
        };
        Timer timer = new Timer("Timer");
        long delay = 1000L * 60;
        long period = 1000L * 60;
        timer.scheduleAtFixedRate(counterTask, delay, period);
    }

    public void incHendelserLest(String tpnr, double antall) {
        try {
            hendelserLestCounterList.get(tpnr).increment(antall);
        } catch (NullPointerException e) {
            LOG.info("No counter for tpnr: " + tpnr);
        }
    }
}

File: src/main/kotlin/no/nav/samordning/hendelser/metrics/AppMetrics.kt
package no.nav.samordning.hendelser.metrics

import io.micrometer.core.instrument.Counter
import io.micrometer.core.instrument.Gauge
import io.micrometer.core.instrument.MeterRegistry
import no.nav.samordning.hendelser.database.HendelseService
import no.nav.samordning.hendelser.ytelse.service.YtelseService
import org.slf4j.LoggerFactory.getLogger
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Bean
import org.springframework.stereotype.Component
import java.util.*

@Component
class AppMetrics(private val registry: MeterRegistry,
    private val hendelseService: HendelseService,
    private val ytelseService: YtelseService) {

    private val log = getLogger(javaClass)

    private var totalAntallHendelser: Number = 0
    private var totalAntallHendelserTpYtelser: Number = 0

    private val hendelserLestCounterList = HashMap<String, Counter>()
    private val hendelserTpYtelserLestCounterList = HashMap<String, Counter>()

    init {
        Gauge.builder("samordning_hendelser_total", ::totalAntallHendelser).register(registry)
        Gauge.builder("samordning_hendelser_tp_ytelser_total", ::totalAntallHendelserTpYtelser).register(registry)
    }

    @Bean
    fun totalHendelserCount() {
        scheduleTask { totalAntallHendelser = hendelseService.totalHendelser }
    }

    @Bean
    fun totalHendelserTpYtelserCount() {
        scheduleTask { totalAntallHendelserTpYtelser = ytelseService.totalHendelsertpYtelser }
    }

    private fun scheduleTask(task: TimerTask.() -> Unit) {
        val counterTask = object : TimerTask() {
            override fun run() {
                task()
            }
        }
        val timer = Timer("Timer")
        val delay = 1000L * 60
        val period = 1000L * 60
        timer.scheduleAtFixedRate(counterTask, delay, period)
    }

    fun incHendelserLest(tpnr: String, antall: Double) {
        try {
            hendelserLestCounterList.getOrPut(tpnr) {
                Counter.builder("samordning_hendelser_lest")
                        .tag("tpnr", tpnr).register(registry)
            }.increment(antall)
        } catch (e: NullPointerException) {
            log.info("No counter for tpnr: $tpnr")
        }
    }

    fun incHendelserTpYtelserLest(tpnr: String, antall: Double) {
        try {
            hendelserTpYtelserLestCounterList.getOrPut(tpnr) {
                Counter.builder("samordning_hendelser_tp_ytelser_lest")
                    .tag("tpnr", tpnr).register(registry)
            }.increment(antall)
        } catch (e: NullPointerException) {
            log.info("No counter for tpnr: $tpnr")
        }
    }
}